Building pyInstaller fails with strange clang error under macOS 10.15 and Homebrew

Hi,

I am having troubles building pyinstaller from Homebrew, since I updated the formula. Among all my other formulas, this is the only one that refuses to build.

I am writing here because it seems to be a pyInstaller issue, not a Homebrew one, and the log says "Do not report this issue to Homebrew/brew or Homebrew/core!"

I think it has to do with the clang C compiler. Maybe a version issue?

I have found a similar error message in issue #1341 [https://github.com/pyinstaller/pyinstaller/issues/1341](1341), but this may be completely unrelated.

The error I get is:

```
[19/20] Linking build/debug/tests/test_launch
Undefined symbols for architecture x86_64:
  "_checkFile", referenced from:
      _test_checkFile in test_launch.c.2.o
  "_splitName", referenced from:
      _test_splitName in test_launch.c.2.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

I have checked the following:

- [X] Homebrew does download the 5.7.0 version (see next section below)
- [X] Unzipping the .gzip, I find that the functions are indeed defined in the c script, as follows:
```int checkFile(char *buf, const char *fmt, ...);
static void test_checkFile(void **state) {
    char result[PATH_MAX];
```
- [X] The installation command seems OK: "python3.11 ./waf all --no-universal2 STRIP=/usr/bin/strip"
- [X] The virtual environment in which this command is executed is managed by Homebrew: I assume everything is OK as all my other formulas build
- [X] I assume that the clang that is called is the one provided by Apple, because I don't have any HomeBrew formula called  'clang' or 'llvm'  ; in the macOS Terminal, the command`which clang` answers:
 ```
Apple clang version 12.0.0 (clang-1200.0.32.29)
Target: x86_64-apple-darwin19.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

### Downloaded files by HomeBrew

```
==> Fetching pyinstaller
==> Downloading https://files.pythonhosted.org/packages/5a/13/a7cfa43856a7b8e4894848ec8f71cd9e1ac461e51802391a3e2101c60ed6/altgraph-0.17.3
Already downloaded: /Users/lagaffe/Library/Caches/Homebrew/downloads/2b32e84161681feb5911d4b89cbecdad53aa6424f06892bfb5cd301a5222cdc5--altgraph-0.17.3.tar.gz
==> Downloading https://files.pythonhosted.org/packages/46/92/bffe4576b383f20995ffb15edccf1c97d2e39f9a8c72136836407f099277/macholib-1.16.2
Already downloaded: /Users/lagaffe/Library/Caches/Homebrew/downloads/107bc2d2781424698f61c7573aea66a182c55487da19f5044fe639987b413989--macholib-1.16.2.tar.gz
==> Downloading https://files.pythonhosted.org/packages/f4/e1/1b4f8394be02574479167b0f09a9fdb3ca4f184b45bd89d5878e44ba72ed/pyinstaller-hoo
Already downloaded: /Users/lagaffe/Library/Caches/Homebrew/downloads/dee33bb1239429c9213ff114d701dcd45fd7037e15a7d2924747f2b3914cb62f--pyinstaller-hooks-contrib-2022.14.tar.gz
==> Downloading https://files.pythonhosted.org/packages/13/2e/bc4c0026659cea46aca867f4d71e8bab5a6430b4c005c51e174da5e6e4a2/pyinstaller-5.7
Already downloaded: /Users/lagaffe/Library/Caches/Homebrew/downloads/77f59edde69eb5fa1a828902612bf1ac26ac41143931cb0097d1d39196f3f625--pyinstaller-5.7.0.tar.gz
```

### Note on my environment

I have to emphasize that I am running MacOS Catalina 10.15.7, so all my formulas have to be built locally instead of using binaries. I have noticed in the logs that pyinstaller assumes that I am running 10.13, but I have managed to force it to use 10.15 by setting MACOSX_DEPLOYMENT_TARGET="10.15" (by modifying the ruby script of the Homebrew formula), and this changes nothing.

(yes, I know that it is no longer supported by Homebrew. Even thoug, could you please not despise me like the Homebrew developers in their disclaimer statement).

### Other log files

The "config.log" and "01.python3.11" log files are available on request.

The most interesting part is the beginning of "01.python3.11":

```
2023-01-18 05:53:25 +0000

python3.11
./waf
all
--no-universal2
STRIP=/usr/bin/strip

'all' finished successfully (0.001s)
'distclean' finished successfully (0.000s)
Setting top to                           : /private/tmp/pyinstaller-20230118-94300-15loc3/pyinstaller-5.7.0/bootloader 
Setting out to                           : /private/tmp/pyinstaller-20230118-94300-15loc3/pyinstaller-5.7.0/bootloader/build 
Python Version                           : 3.11.1 (main, Jan 15 2023, 20:19:46) [Clang 12.0.0 (clang-1200.0.32.29)] 
Checking for 'clang' (C compiler)        : clang 
MacOS deployment target                  : Setting to 10.13 
Checking size of pointer                 : 8 
Platform                                 : Darwin-64bit detected based on compiler 
Checking for library m                   : yes 
Checking for library z                   : yes 
Checking for library cmocka              : yes 
Checking for function unsetenv           : yes 
Checking for function mkdtemp            : no 
Checking for function dirname            : yes 
Checking for function basename           : yes 
Checking for function strndup            : yes 
Checking for function strnlen            : yes 
Checking for program 'strip'             : /usr/bin/strip 
Checking for program 'strip'             : /usr/bin/strip 
'configure' finished successfully (2.357s)
'make_all' finished successfully (0.004s)
Waf: Entering directory `/private/tmp/pyinstaller-20230118-94300-15loc3/pyinstaller-5.7.0/bootloader/build/debug'
[ 1/20] Compiling src/pyi_path.c
[ 2/20] Compiling src/pyi_exception_dialog.c
[ 3/20] Compiling src/pyi_apple_events.c
[ 4/20] Compiling src/pyi_win32_utils.c
[ 5/20] Compiling src/pyi_main.c
[ 6/20] Compiling src/pyi_utils.c
[ 7/20] Compiling src/pyi_launch.c
[ 8/20] Compiling src/pyi_python.c
[ 9/20] Compiling src/pyi_global.c
[10/20] Compiling src/pyi_splashlib.c
[11/20] Compiling src/pyi_archive.c
[12/20] Compiling src/pyi_pythonlib.c
[13/20] Compiling src/pyi_splash.c
[14/20] Compiling src/main.c
[15/20] Compiling tests/test_path.c
[16/20] Compiling tests/test_launch.c
[17/20] Linking build/debug/run_d
[18/20] Linking build/debug/tests/test_path
[19/20] Linking build/debug/tests/test_launch
ld: warning: dylib (/usr/local/lib/libcmocka.dylib) was built for newer macOS version (10.15) than being linked (10.13)

ld: warning: dylib (/usr/local/lib/libcmocka.dylib) was built for newer macOS version (10.15) than being linked (10.13)
Undefined symbols for architecture x86_64:
  "_checkFile", referenced from:
      _test_checkFile in test_launch.c.2.o
  "_splitName", referenced from:
      _test_splitName in test_launch.c.2.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)

Waf: Leaving directory `/private/tmp/pyinstaller-20230118-94300-15loc3/pyinstaller-5.7.0/bootloader/build/debug'
Build failed
 -> task in 'test_launch' failed with exit status 1 (run with -v to display more information)
```

Thank you

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -46,6 +46,7 @@ jobs:
         run: |
           sudo apt-get update -qq
           sudo apt-get install -qq --no-install-recommends \
+            libcmocka-dev \
             libxml2-dev libxslt1-dev gfortran libatlas-base-dev \
             libespeak1 libxcb-image0 libxcb-keysyms1 libxcb-render-util0 \
             libxkbcommon-x11-0 libxcb-icccm4 libxcb1 openssl \

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -81,17 +81,17 @@ jobs:
         run: |
           # Compile bootloader
           cd bootloader
-          CC="gcc -std=gnu90" python waf all
+          CC="gcc -std=gnu90" python waf --tests all
 
       - name: Check if bootloader is buildable with --static-zlib option
         if: startsWith(matrix.os, 'ubuntu')
-        run: cd bootloader && python waf --static-zlib all
+        run: cd bootloader && python waf --static-zlib --tests all
 
       - name: Check if bootloader is buildable for Windows ARM
         if: startsWith(matrix.os, 'windows')
         run: |
             cd bootloader
-            python waf --target-arch=64bit-arm all
+            python waf --tests --target-arch=64bit-arm all
             ls ../PyInstaller/bootloader/Windows-64bit-arm
 
       - name: Install PyInstaller
@@ -101,7 +101,7 @@ jobs:
 
           # Compile bootloader
           cd bootloader
-          python waf all
+          python waf --tests all
           cd ..
 
           # Install PyInstaller.

--- a/bootloader/src/pyi_launch.c
+++ b/bootloader/src/pyi_launch.c
@@ -53,7 +53,8 @@
  */
 
 /* Constructs the file path from given components and checks that the path exists. */
-static int
+/* NOTE: must be visible outside of this unit (i.e., non-static) due to tests! */
+int
 _format_and_check_path(char *buf, const char *fmt, ...)
 {
     va_list args;
@@ -69,7 +70,8 @@ _format_and_check_path(char *buf, const char *fmt, ...)
 }
 
 /* Splits the item in the form path:filename */
-static int
+/* NOTE: must be visible outside of this unit (i.e., non-static) due to tests! */
+int
 _split_dependency_name(char *path, char *filename, const char *item)
 {
     char *p;

--- a/bootloader/tests/test_launch.c
+++ b/bootloader/tests/test_launch.c
@@ -20,20 +20,20 @@
 #include <setjmp.h> // required fo cmocka :-(
 #include <cmocka.h>
 
-int checkFile(char *buf, const char *fmt, ...);
+int _format_and_check_path(char *buf, const char *fmt, ...);
 
-static void test_checkFile(void **state) {
+static void test__format_and_check_path(void **state) {
     char result[PATH_MAX];
 
     // TODO: use some mocks to determine stat() output
 
     errno = 0;
-    assert_int_equal(-1, checkFile(result, "%s%s%s.pkg", "a1", "bb", "cc", "dd"));
+    assert_int_equal(-1, _format_and_check_path(result, "%s%s%s.pkg", "a1", "bb", "cc", "dd"));
     assert_int_not_equal(errno, 0); // formatting passed, stat failed
     assert_string_equal(result, "a1bbcc.pkg");
 
     errno = 0;
-    assert_int_equal(-1, checkFile(result, "%s", ""));
+    assert_int_equal(-1, _format_and_check_path(result, "%s", ""));
     assert_int_not_equal(errno, 0); // formatting passed, stat failed
     assert_string_equal(result, "");
 
@@ -42,55 +42,55 @@ static void test_checkFile(void **state) {
     // a few bytes more
     errno = 0;
     path2[PATH_MAX+8] = '\0';
-    assert_int_equal(-1, checkFile(result, "%s%s%s.pkg", "a1", path2, "ccc"));
+    assert_int_equal(-1, _format_and_check_path(result, "%s%s%s.pkg", "a1", path2, "ccc"));
     assert_int_equal(errno, 0); // formatting formatting failed
     // exact length
     errno = 0;
     path2[PATH_MAX] = '\0';
-    assert_int_equal(-1, checkFile(result, "%s", path2));
+    assert_int_equal(-1, _format_and_check_path(result, "%s", path2));
     assert_int_equal(errno, 0); // formatting formatting failed
     // one byte less
     errno = 0;
     path2[PATH_MAX-1] = '\0';
-    assert_int_equal(-1, checkFile(result, "%s", path2));
+    assert_int_equal(-1, _format_and_check_path(result, "%s", path2));
     assert_int_not_equal(errno, 0); // formatting passed, stat failed
 }
 
-int splitName(char *path, char *filename, const char *item);
+int _split_dependency_name(char *path, char *filename, const char *item);
 
-static void test_splitName(void **state) {
+static void test_split_dependency_name(void **state) {
     char path[PATH_MAX];
     char filename[PATH_MAX];
 
     // TODO: use some mocks to determine
 
-    assert_int_equal(0, splitName(path, filename, "aaa:bbb"));
+    assert_int_equal(0, _split_dependency_name(path, filename, "aaa:bbb"));
     assert_string_equal(path, "aaa");
     assert_string_equal(filename, "bbb");
 
-    assert_int_equal(-1, splitName(path, filename, ""));
-    assert_int_equal(-1, splitName(path, filename, ":"));
-    assert_int_equal(-1, splitName(path, filename, "aaa"));
-    assert_int_equal(-1, splitName(path, filename, "aaa:"));
-    assert_int_equal(-1, splitName(path, filename, ":bbb"));
+    assert_int_equal(-1, _split_dependency_name(path, filename, ""));
+    assert_int_equal(-1, _split_dependency_name(path, filename, ":"));
+    assert_int_equal(-1, _split_dependency_name(path, filename, "aaa"));
+    assert_int_equal(-1, _split_dependency_name(path, filename, "aaa:"));
+    assert_int_equal(-1, _split_dependency_name(path, filename, ":bbb"));
 
     // these cases are not expected to occur in real life
-    assert_int_equal(0, splitName(path, filename, "aaa:::"));
+    assert_int_equal(0, _split_dependency_name(path, filename, "aaa:::"));
     assert_string_equal(filename, "::");
-    assert_int_equal(-1, splitName(path, filename, ":::bbb"));
+    assert_int_equal(-1, _split_dependency_name(path, filename, ":::bbb"));
 
     char *path2 = (char *) malloc(PATH_MAX+10);
     memset(path2, 'a', PATH_MAX+8);
     path2[10] = ':';
     // a few bytes more
     path2[PATH_MAX+8] = '\0';
-    assert_int_equal(-1, splitName(path, filename, path2));
+    assert_int_equal(-1, _split_dependency_name(path, filename, path2));
     // exact length
     path2[PATH_MAX] = '\0';
-    assert_int_equal(-1, splitName(path, filename, path2));
+    assert_int_equal(-1, _split_dependency_name(path, filename, path2));
     // one byte less
     path2[PATH_MAX-1] = '\0';
-    assert_int_equal(0, splitName(path, filename, path2));
+    assert_int_equal(0, _split_dependency_name(path, filename, path2));
     assert_string_equal(path, "aaaaaaaaaa");
 }
 
@@ -101,8 +101,8 @@ int main(void)
 #endif
 {
     const struct CMUnitTest tests[] = {
-        cmocka_unit_test(test_checkFile),
-        cmocka_unit_test(test_splitName),
+        cmocka_unit_test(test__format_and_check_path),
+        cmocka_unit_test(test_split_dependency_name),
     };
     return cmocka_run_group_tests(tests, NULL, NULL);
 }

--- a/bootloader/tests/wscript
+++ b/bootloader/tests/wscript
@@ -19,9 +19,9 @@ def build(ctx):
         if ctx.env.DEST_OS == 'win32':
             # Z: inflate*()
             # ADVAPI32: ConvertStringSecurityDescriptorToSecurityDescriptorW()
-            extra_libs=['ADVAPI32', 'Z']
+            extra_libs=['ADVAPI32', 'Z', 'STATIC_ZLIB']
         else:
-            extra_libs=[]
+            extra_libs=['STATIC_ZLIB']
         ctx.program(
             source= ["test_%s.c" % name],
             target="test_%s" % name,

--- a/bootloader/tests/wscript
+++ b/bootloader/tests/wscript
@@ -11,7 +11,10 @@
 # -----------------------------------------------------------------------------
 
 def configure(ctx):
-    ctx.check_cc(lib='cmocka', mandatory=False, uselib_store='CMOCKA')
+    ctx.msg('Build tests', "enabled" if ctx.options.enable_tests else "disabled")
+    if ctx.options.enable_tests:
+        ctx.check_cc(lib='cmocka', mandatory=False, uselib_store='CMOCKA')
+
 
 def build(ctx):
 
@@ -37,6 +40,6 @@ def build(ctx):
         # be WinMain() instead of main().
         return
 
-    if "LIB_CMOCKA" in ctx.env:
+    if ctx.options.enable_tests and "LIB_CMOCKA" in ctx.env:
         test_program("path")
         test_program("launch")

--- a/bootloader/wscript
+++ b/bootloader/wscript
@@ -170,6 +170,20 @@ def options(ctx):
         'This is always done on Windows.',
         default=False,
     )
+    ctx.add_option(
+        '--tests',
+        action='store_true',
+        help='Enable cmocka-based tests if cmocka library is found. The tests are disabled by default.',
+        default=False,
+        dest='enable_tests',
+    )
+    ctx.add_option(
+        '--no-tests',
+        action='store_false',
+        help='Disable cmocka-based tests, even if cmocka library is found. The tests are disabled by default.',
+        default=False,
+        dest='enable_tests',
+    )
 
     grp = ctx.add_option_group('Linux Standard Base (LSB) compliance', 'These options have effect only on Linux.')
     grp.add_option(

